mitmproxy/addons/dns_resolver.py

```diff
@@ -155,7 +67,56 @@ async def dns_request(self, flow: dns.DNSFlow) -> None:
             and not flow.error
         )
         if should_resolve:
-            # TODO: We need to handle overly long responses here.
-            flow.response = await resolve_message(
-                flow.request, asyncio.get_running_loop()
+            all_ip_lookups = (
+                flow.request.query
+                and flow.request.op_code == dns.op_codes.QUERY
+                and all(
+                    q.type in (dns.types.A, dns.types.AAAA)
+                    and q.class_ == dns.classes.IN
+                    for q in flow.request.questions
+                )
             )
+            # We use `mitmproxy_rs.DnsResolver` if we need to use the hosts file to lookup hostnames(A/AAAA queries only)
+            # For other cases we forward it to the specified name server directly.
+            if all_ip_lookups and ctx.options.use_hosts_file:
+                # TODO: We need to handle overly long responses here.
+                flow.response = await self.resolve_message(flow.request)
+            elif not flow.server_conn.address:
+                flow.server_conn.address = (self.name_servers[0], 53)
+
+    async def resolve_message(self, message: dns.Message) -> dns.Message:
+        try:
+            rrs: list[dns.ResourceRecord] = []
+            for question in message.questions:
+                rrs.extend(await self.resolve_question(question))
+        except ResolveError as e:
+            return message.fail(e.response_code)
+        else:
+            return message.succeed(rrs)
+
+    async def resolve_question(
+        self, question: dns.Question
+    ) -> Iterable[dns.ResourceRecord]:
+        assert question.type in (dns.types.A, dns.types.AAAA)
+
+        try:
+            if question.type == dns.types.A:
+                addrinfos = await self.resolver.lookup_ipv4(question.name)
+            elif question.type == dns.types.AAAA:
+                addrinfos = await self.resolver.lookup_ipv6(question.name)
+        except socket.gaierror as e:
+            if e.args[0] == "NXDOMAIN":
+                raise ResolveError(dns.response_codes.NXDOMAIN)
+            else:  # pragma: no cover
+                raise ResolveError(dns.response_codes.SERVFAIL)
```

Comment:
How do we want to handle `NOERROR` codes here i.e `if e.args[0] == "NOERROR"`? 
(we can get a `NOERROR` here if there are no records returned but the name is valid)
https://datatracker.ietf.org/doc/html/rfc2308#section-2.2

Reply:
Then we should also send a NOERROR to the client, no?

Reply:
What about the authority section? Can we ignore that?

Reply:
Yes, I'd say let's just ignore that for now. :)